src/devices/Arduino/FirmataDevice.cs

```diff
@@ -1074,23 +1077,53 @@ public void WriteReadI2cData(int slaveAddress,  ReadOnlySpan<byte> writeData, Sp
                 {
                     throw new IOException($"Expected {replyData.Length} bytes, got only {bytesReceived}");
                 }
+
+                _numberOfConsecutiveI2cWrites = 0; // after we get a reply, we can free-fire a few times again
             }
             else
             {
                 SendCommand(i2cSequence);
+                // If we use a series of write-only I2C commands we have to occasionally introduce a command that requires an answer, or we flood the device's input buffer,
+                // causing network retries, which under the line causes more delays than this. This problem is particularly obvious with an ESP32 in Wifi mode
+                _numberOfConsecutiveI2cWrites++;
+                if (_numberOfConsecutiveI2cWrites % 4 == 0)
+                {
+                    var pin = _supportedPinConfigurations.FirstOrDefault(x => x.PinModes.Contains(SupportedMode.I2c));
+                    if (pin != null)
+                    {
+                        GetPinMode(pin.Pin);
+                    }
+                    else
+                    {
+                        Thread.Sleep(10); // Hopefully, this doesn't happen
+                    }
+                }
             }
         }
 
         public void SetPwmChannel(int pin, double dutyCycle)
         {
-            FirmataCommandSequence pwmCommandSequence = new();
-            pwmCommandSequence.WriteByte((byte)FirmataSysexCommand.EXTENDED_ANALOG);
-            pwmCommandSequence.WriteByte((byte)pin);
             // The arduino expects values between 0 and 255 for PWM channels.
-            // The frequency cannot be set.
+            // The frequency cannot currently be set using the protocol
             int pwmMaxValue = _supportedPinConfigurations[pin].PwmResolutionBits; // This is 8 for most arduino boards
             pwmMaxValue = (1 << pwmMaxValue) - 1;
             int value = (int)Math.Max(0, Math.Min(dutyCycle * pwmMaxValue, pwmMaxValue));
+
+            // At most 14 bits used?
+            if (((pwmMaxValue & 0x3FFF) == pwmMaxValue) && (pin <= 15))
```

Comment:
same here, why the 15 pins limits?

Reply:
See other comment: If the pin number is <=15, a shorthand form of the command can be used.